### PR TITLE
[TASK] Deprecate support for RequireJS modules

### PR DESCRIPTION
**Pull request by linawolf**
resolves https://github.com/TYPO3-Documentation/Changelog-To-Doc/issues/216
releases: main

### CODE REVIEW

File: Documentation/ApiOverview/Backend/BroadcastChannels.rst [hunk 1]
```diff
@@ -75,29 +75,32 @@ Example code:
    });
 
 
-Hook into :php:`$GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['typo3/backend.php']['constructPostProcess']` to load a custom
-:php:`BackendController` hook that loads the event handler, e.g. via RequireJS.
+Hook into
+:php:`$GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['typo3/backend.php']['constructPostProcess']`
+to load a custom :php:`BackendController` hook that loads the event handlers
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/BroadcastChannels.rst:80]:
```suggestion
to load a custom :php:`BackendController` hook that loads the event handler's
```

Above is a `require(['TYPO3/CMS/Backend/BroadcastService']` line. Just add a todo that this has to be changed the also (or do it now, if applicable) - or it is a completely separate task?

File: Documentation/ApiOverview/Backend/BroadcastChannels.rst [hunk 2]
```diff
@@ -75,29 +75,32 @@ Example code:
...
+
+    use TYPO3\CMS\Core\Utility\GeneralUtility;
+    use TYPO3\CMS\Core\Page\PageRenderer;
+
+    class BackendControllerHook
+    {
+        public function registerClientSideEventHandler(): void
+        {
+            $pageRenderer = GeneralUtility::makeInstance(PageRenderer::class);
+            $pageRenderer->loadJavaScriptModule('@myvendor/my-extension/event-handler.js');
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/BroadcastChannels.rst:103]: Do we have a page which describes that JavaScript ES6 stuff (@myvendor/my-extension/event-handler.js)? So I propose to link to it in a seealso or something to guide the user what it is all about.

File: Documentation/ApiOverview/Backend/BroadcastChannels.rst [hunk 3]
```diff
@@ -75,29 +75,32 @@ Example code:
...
-       {
-           $pageRenderer = GeneralUtility::makeInstance(PageRenderer::class);
-           $pageRenderer->loadRequireJsModule('TYPO3/CMS/MyExtension/EventHandler');
-       }
+    :caption: EXT:my_extension/Classes/Hooks/BackendControllerHook.php
+
+    use TYPO3\CMS\Core\Utility\GeneralUtility;
+    use TYPO3\CMS\Core\Page\PageRenderer;
+
+    class BackendControllerHook
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/BroadcastChannels.rst:98]:
```suggestion
    final class BackendControllerHook
```

You know, I couldn't resist ;-)

File: Documentation/ApiOverview/Backend/ContextualMenu.rst [hunk 4]
```diff
@@ -31,15 +22,18 @@ Context menu rendering flow
...
 
-The context menu is shown after click on the HTML element which has
+The context menu is shown after clicking on the HTML element which has
 :html:`class="t3js-contextmenutrigger"` together with :html:`data-table`,
 :html:`data-uid` and optional :html:`data-context` attributes.
 
 The JavaScript click event handler is implemented in the
-:js:`TYPO3/CMS/Backend/ContextMenu` requireJS module. It takes the
-data attributes mentioned above and executes an ajax call to the
+ES6 module :js:`@typo3/backend/context-menu.js`. It takes the
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/ContextualMenu.rst:30]: Also here, can we link "ES6 module" to a page which explains it? For me, I would be lost now.

File: Documentation/ApiOverview/Backend/ContextualMenu.rst [hunk 5]
```diff
@@ -31,15 +22,18 @@ Context menu rendering flow
...
+data attributes mentioned above and executes an AJAX call to the
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/ContextualMenu.rst:31]: Wikipedia always writes it "Ajax" also mentioning it my be written "AJAX". It think it is getting just like a word, as the X does not stand anymore for XML but can also by JSON or whatever:
https://en.wikipedia.org/wiki/Ajax_(programming)

File: Documentation/ApiOverview/Backend/ContextualMenu.rst [hunk 6]
```diff
@@ -31,15 +22,18 @@ Context menu rendering flow
...
 :php:`\TYPO3\CMS\Backend\Controller\ContextMenuController->getContextMenuAction()`.
 
+..  versionchanged:: 12.0
+    The RequireJS module :js:`TYPO3/CMS/Backend/ContextMenu` has been migrated
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/ContextualMenu.rst:35]: What does that mean to me? What have I to do? Can we link to the tutorial section below (although it is not prepared for ES6 yet)?

File: Documentation/ApiOverview/Backend/ContextualMenu.rst [hunk 7]
```diff
@@ -94,11 +89,14 @@ After that, a compiled list of items is returned to the
 Menu rendering in JavaScript
 ----------------------------
 
-Based on the JSON data :file:`ContextMenu.js` is rendering a context menu. If
+..  versionchanged:: 12.0
+    The RequireJS module :js:`TYPO3/CMS/Backend/ContextMenuActions` has been migrated
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/ContextualMenu.rst:93]: What does this mean to me?

File: Documentation/ApiOverview/Backend/JavaScript/RequireJS/Dependency/Index.rst [hunk 8]
```diff
@@ -6,6 +6,15 @@
 Dependency handling
 ===================
 
+.. deprecated::  12.0
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/RequireJS/Dependency/Index.rst:9]:
```suggestion
..  deprecated::  12.0
```

File: Documentation/ApiOverview/Backend/JavaScript/RequireJS/Dependency/Index.rst [hunk 9]
```diff
@@ -6,6 +6,15 @@
...
+    The RequireJS project has been discontinued_ and was therefore
+    replaced by native ECMAScript v6/v11 modules in TYPO3 v12.0. The
+    infrastructure for configuration and loading of RequireJS
+    modules is deprecated with v12.0 and will be removed in TYPO3 v13. See
+    :ref:`RequireJS to ES6 migration <requirejs-migration>`.
+
+.. _discontinued: https://github.com/requirejs/requirejs/issues/1816
+
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/RequireJS/Dependency/Index.rst:17]: Should be add a hint that the following code it outdated by now?

File: Documentation/ApiOverview/Backend/JavaScript/RequireJS/Extensions/Index.rst [hunk 10]
```diff
@@ -6,6 +6,15 @@
 Use RequireJS in your own extension
 ===================================
 
+.. deprecated::  12.0
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/RequireJS/Extensions/Index.rst:9]:
```suggestion
..  deprecated::  12.0
```

This should be also made more prominent, maybe wrap in an attention admonition.

File: Documentation/ApiOverview/Backend/JavaScript/RequireJS/Index.rst [hunk 11]
```diff
@@ -6,27 +6,114 @@
 RequireJS in the TYPO3 Backend
 ==============================
 
-Since TYPO3 CMS 7 it is possible to use RequireJS in the backend.
-A short explanation: RequireJS enables developers to have some kind of
-dependency handling for JavaScript. The JavaScript is written as so-called
-"Asynchronous Module Definition" (AMD).
-Some libraries delivered with TYPO3 are written as modules.
+.. deprecated::  12.0
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/RequireJS/Index.rst:9]:
```suggestion
..  deprecated::  12.0
```

This should be also made more prominent, maybe wrap in an attention admonition.

File: Documentation/ApiOverview/Backend/JavaScript/RequireJS/Index.rst [hunk 12]
```diff
@@ -6,27 +6,114 @@
...
+work in TYPO3 v12. These modules will be loaded after modules registered via
+:php:`$pageRenderer->javaScriptModules`. Extensions that use
+:php:`$pageRenderer->requireJsModules` will work as before but trigger a
+PHP :php:`E_USER_DEPRECATED` error.
+
+If your extension wants to support both TYPO3 v11 and v12 you can keep the
+RequireJS version and remove it upon switching to TYPO3 v13.
+
+If you want to prevent deprecation warnings you can also implement both
+RequireJS (with a version switch) and a native ECMAScript v6/v11 (ES6) modules.
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/RequireJS/Index.rst:39]:
```suggestion
RequireJS (with a version switch) and native ECMAScript v6/v11 (ES6) modules.
```

File: Documentation/ApiOverview/Backend/JavaScript/RequireJS/Index.rst [hunk 13]
```diff
@@ -6,27 +6,114 @@
...
+This approach is recommended if you are working with TypeScript and the
+JavaScript will be generated anyway.
+
+..  todo: Set link to ES6 description instead of issue once
+    https://github.com/TYPO3-Documentation/Changelog-To-Doc/issues/5 is resolved
+
+Migrate your JavaScript from the AMD module format to native ES6 modules
+and register your configuration in
+:file:`EXT:my_extension/Configuration/JavaScriptModules.php`, also see
+:issue:`96510` for more information:
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/RequireJS/Index.rst:49]: What value has the link to the issue? I see only a list of four items in the description?

File: Documentation/ApiOverview/Backend/JavaScript/RequireJS/Index.rst [hunk 14]
```diff
@@ -6,27 +6,114 @@
...
+    :caption: EXT:my_extension/Classes/SomeClass.php
 
-   Extensions/Index
-   Dependency/Index
-   Loading/Index
-   Shim/Index
+    use TYPO3\CMS\Core\Information\Typo3Version;
+    use TYPO3\CMS\Core\Page\PageRenderer;
+
+    $this->pageRenderer->loadJavaScriptModule(
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/RequireJS/Index.rst:73]: The `loadJavaScriptModule()` method was introduced with v12, right? So there will be a PHP error on v11, I assume?

So:
TYPO3 >= 12: use loadJavaScriptModule()
else: use loadRequireJsModule()

This way you can also have an extension compatible with v11-v13 ;-)

File: Documentation/ApiOverview/Backend/JavaScript/RequireJS/Index.rst [hunk 15]
```diff
@@ -6,27 +6,114 @@
...
+        // keep RequireJs for TYPO3 below v12.0
+        $this->pageRenderer->loadRequireJsModule(
+            'TYPO3/CMS/MyExtension/MyExample'
+        );
+    }
+
+In Fluid templates `includeJavaScriptModules` is to be used instead of
+`includeRequireJsModules`.
+
+In Fluid template the `includeJavaScriptModules` property of the
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/RequireJS/Index.rst:88]:
```suggestion
In Fluid template the `includeJavaScriptModules` attribute of the
```
as we are in XML.

This sentence and above are similar.

File: Documentation/ApiOverview/FormEngine/Rendering/Index.rst [hunk 16]
```diff
@@ -190,19 +190,24 @@ of a child can be slightly different. For instance, a :php:`FieldControl` "wizar
...
+:php:`$resultArray = $this->initializeResultArray()` takes care of basic keys:
 
 .. code-block:: php
 
     [
         'html' => '',
         'additionalInlineLanguageLabelFiles' => [],
         'stylesheetFiles' => [],
+        'javaScriptModules' => $javaScriptModules,
+        /** @deprecated will be removed in TYPO3 v13.0 */
```
Comment (brotkrueml) [inline Documentation/ApiOverview/FormEngine/Rendering/Index.rst:203]:
```suggestion
        /** @deprecated requireJsModules will be removed in TYPO3 v13.0 */
```

To make it really clear that the below is meant.

File: Documentation/ApiOverview/Icon/Index.rst [hunk 17]
```diff
@@ -116,72 +116,95 @@ This will render the desired icon using an `img`-tag. If you prefer having the S
 The JavaScript way
 ------------------
 
+..  versionchanged:: 12.0
+    The JavaScript icon provider has been moved from the RequireJS module
+    :js:`TYPO3/CMS/Backend/Icons` to the ES6 module :js:``
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Icon/Index.rst:121]:
```suggestion
    :js:`TYPO3/CMS/Backend/Icons` to the ES6 module :js:`@typo3/backend/icons`.
```

File: Documentation/ApiOverview/Backend/BroadcastChannels.rst (hunk 2 above)
Comment (linawolf) [reply Documentation/ApiOverview/Backend/BroadcastChannels.rst:103]: Nope, not yet, still working on it: https://github.com/TYPO3-Documentation/Changelog-To-Doc/issues/5

File: Documentation/ApiOverview/Backend/ContextualMenu.rst (hunk 7 above)
Comment (linawolf) [reply Documentation/ApiOverview/Backend/ContextualMenu.rst:93]: Well it is like with switching from Hooks to Events. The one thing is to learn to apply the new technique, The other thing is that there are new names for all the API. In these comments I I mention the old names once more just like we do mention the old hook when it is replaced by an event?

File: Documentation/ApiOverview/Backend/JavaScript/RequireJS/Dependency/Index.rst (hunk 9 above)
Comment (linawolf) [reply Documentation/ApiOverview/Backend/JavaScript/RequireJS/Dependency/Index.rst:17]: Well the old code still works but causes deprecation warnings in v12

File: Documentation/ApiOverview/Backend/BroadcastChannels.rst (hunk 2 above)
Comment (brotkrueml) [reply Documentation/ApiOverview/Backend/BroadcastChannels.rst:103]: I appreciate your work on that JS stuff, for me it is still hard to grab and understand - thumbs up from my side.
So, maybe just add a todo, so someone(TM) will stumble upon it at someday and adds the link :)

File: Documentation/ApiOverview/Backend/JavaScript/RequireJS/Index.rst (hunk 13 above)
Comment (linawolf) [reply Documentation/ApiOverview/Backend/JavaScript/RequireJS/Index.rst:49]: The naming of the ticket is maybe not so good, if you follow to the Gerrit Change, the title is better: https://review.typo3.org/c/Packages/TYPO3.CMS/+/73004

File: Documentation/ApiOverview/Backend/JavaScript/RequireJS/Index.rst [hunk 18]
```diff
@@ -6,27 +6,114 @@
...
+    use TYPO3\CMS\Core\Page\PageRenderer;
+
+    $this->pageRenderer->loadJavaScriptModule(
+        '@vendor/my-extension/my-example.js'
+    );
+
+    $typo3Version = new Typo3Version();
+    if ($typo3Version->getMajorVersion() < 12) {
+        // keep RequireJs for TYPO3 below v12.0
+        $this->pageRenderer->loadRequireJsModule(
```
Comment (linawolf) [inline Documentation/ApiOverview/Backend/JavaScript/RequireJS/Index.rst:80]:
```suggestion
    use TYPO3\CMS\Core\Page\PageRenderer;

    $typo3Version = new Typo3Version();
    if ($typo3Version->getMajorVersion() > 11) {
        $this->pageRenderer->loadJavaScriptModule(
            '@vendor/my-extension/my-example.js'
        );
    } else {
        // keep RequireJs for TYPO3 below v12.0
        $this->pageRenderer->loadRequireJsModule(
```

File: Documentation/ApiOverview/Backend/BroadcastChannels.rst (hunk 2 above)
Comment (linawolf) [reply Documentation/ApiOverview/Backend/BroadcastChannels.rst:103]: How about we make a page stub where we add a todo to explain ES6 and then link there from all the places that need links?

File: Documentation/ApiOverview/Backend/ContextualMenu.rst (hunk 5 above)
Comment (linawolf) [inline Documentation/ApiOverview/Backend/ContextualMenu.rst:31]:
```suggestion
data attributes mentioned above and executes an Ajax call to the
```

File: Documentation/ApiOverview/Backend/ContextualMenu.rst (hunk 4 above)
Comment (linawolf) [reply Documentation/ApiOverview/Backend/ContextualMenu.rst:30]: From what I understand "ES6" is a common concept. There must be sources on this somewhere (TM) but I am just starting to learn about it myself. And the guys who made the changes didn't spend much time on documentation

File: Documentation/ApiOverview/Backend/BroadcastChannels.rst (hunk 2 above)
Comment (brotkrueml) [reply Documentation/ApiOverview/Backend/BroadcastChannels.rst:103]: Thumbs up from me :-) It is main, so WIP. And perhaps someone who reads it want to contribute (dream).

Can we add something like: If you know about this topic please consider to contribute. 8with link to the contribution page).

Comment (linawolf) [reply Documentation/ApiOverview/Backend/BroadcastChannels.rst:103]: I actually already assigned myself to the issue. Contributions would be a dream but it doesn't seem like any are coming for now

File: Documentation/ApiOverview/Backend/ContextualMenu.rst (hunk 7 above)
Comment (linawolf) [reply Documentation/ApiOverview/Backend/ContextualMenu.rst:94]: added a link

File: Documentation/ApiOverview/Backend/JavaScript/RequireJS/Dependency/Index.rst (hunk 9 above)
Comment (linawolf) [reply Documentation/ApiOverview/Backend/JavaScript/RequireJS/Dependency/Index.rst:18]: with the attention box it should become clearer

File: Documentation/ApiOverview/Backend/BroadcastChannels.rst (hunk 2 above)
Comment (linawolf) [reply Documentation/ApiOverview/Backend/BroadcastChannels.rst:123]: done

File: Documentation/ApiOverview/Backend/ContextualMenu.rst (hunk 6 above)
Comment (linawolf) [reply Documentation/ApiOverview/Backend/ContextualMenu.rst:35]: added link

File: Documentation/ApiOverview/Backend/ContextualMenu.rst (hunk 4 above)
Comment (linawolf) [reply Documentation/ApiOverview/Backend/ContextualMenu.rst]: added links

File: Documentation/ApiOverview/Backend/JavaScript/RequireJS/Index.rst (hunk 13 above)
Comment (linawolf) [reply Documentation/ApiOverview/Backend/JavaScript/RequireJS/Index.rst]: linked to the stub instead

File: Documentation/ApiOverview/Backend/BroadcastChannels.rst [hunk 19]
```diff
@@ -29,26 +38,30 @@ component name and the event name, e.g. `typo3:my_extension:my_event`.
 
 .. attention::
 
-   Since a polyfill is in place to add support for Microsoft Edge, the payload must contain JSON-serializable content
-   only.
-
+    Since a polyfill is in place to add support for Microsoft Edge, the
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/BroadcastChannels.rst:41]: This admonition can be deleted. The polyfill was assumably for older Edge versions which are not Chromium-based. The polyfill was removed with commit https://github.com/TYPO3/typo3/commit/47481cbdd8779f15d69095ef479156884e522929#diff-ce8f178fabf4fc619e29c4a9a6caabdc90f8769fc10d631dd1aba54d24801349.

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst [hunk 20]
```diff
@@ -9,10 +9,20 @@ ES6 in the TYPO3 Backend
...
+    *   :ref:`ext_core:feature-96510`
+    *   :ref:`ext_core:feature-98431-1664652179`
+
+
+.. _backend-javascript-es6-loading:
+
+Loading ES6
+===========
+
+..  info::
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:24]: The admonition is not rendered. Strange.

File: Documentation/ApiOverview/Backend/BroadcastChannels.rst [hunk 21]
```diff
@@ -18,8 +19,16 @@ It is possible to send broadcast messages from anywhere in TYPO3 that are listen
 Send a message
 --------------
 
-Any backend module may send a message using the :js:`TYPO3/CMS/Backend/BroadcastService` module.
-The payload of such message is an object that consists at least of the following properties:
+..  versionchanged:: 12.0
+    The RequireJS module :js:`TYPO3/CMS/Backend/BroadcastService` has been migrated
+    to the ES6 module :js:`@typo3/backend/broadcast-service.js.
```
Comment (brotkrueml) [inline Documentation/ApiOverview/Backend/BroadcastChannels.rst:24]:
```suggestion
    to the ES6 module :js:`@typo3/backend/broadcast-service.js`.
```

File: Documentation/ApiOverview/FormEngine/Rendering/Index.rst [hunk 22]
```diff
@@ -213,39 +218,83 @@ CSS and language labels (which can be used in JS) are added with their file name
...
-------------------------
-
-.. deprecated:: 11.5
-   Using callback functions is deprecated and shall be replaced with new
-   JavaScriptModuleInstruction declarations. In FormEngine, loading the RequireJS
-   module via arrays is deprecated and has to be migrated as well.
+Adding JavaScript modules
+-------------------------
 
 JavaScript is added via RequireJS modules using the
```
Comment (brotkrueml) [inline Documentation/ApiOverview/FormEngine/Rendering/Index.rst:224]: Here is a reference to RequireJS, this should be adapted to be "ES6 modules"?

**Review by brotkrueml — CHANGES_REQUESTED** [review]

File: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst (hunk 20 above)
Comment (linawolf) [reply Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst:24]: There is no info admonition. I always forget that

**Comment by linawolf** [comment]
@brotkrueml can we merge this or is something still missing?

**Review by brotkrueml — APPROVED** [review]

**Pull request merged**